libraries/Microsoft.Bot.Builder.Azure/CosmosDbPartitionedStorage.cs

```diff
@@ -0,0 +1,256 @@
+            {
+                throw new ArgumentException("AuthKey for CosmosDB is required.", nameof(cosmosDbStorageOptions.AuthKey));
+            }
+
+            if (string.IsNullOrEmpty(cosmosDbStorageOptions.DatabaseId))
+            {
+                throw new ArgumentException("DatabaseId is required.", nameof(cosmosDbStorageOptions.DatabaseId));
+            }
+
+            if (string.IsNullOrEmpty(cosmosDbStorageOptions.ContainerId))
+            {
+                throw new ArgumentException("ContainerId is required.", nameof(cosmosDbStorageOptions.ContainerId));
+            }
+
+            _cosmosDbStorageOptions = cosmosDbStorageOptions;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CosmosDbPartitionedStorage"/> class.
+        /// using the provided CosmosDB credentials, database ID, and collection ID.
+        /// </summary>
+        /// <param name="cosmosDbStorageOptions">Cosmos DB partitioned storage configuration options.</param>
+        /// <param name="jsonSerializer">If passing in a custom JsonSerializer, we recommend the following settings:
+        /// <para>jsonSerializer.TypeNameHandling = TypeNameHandling.All.</para>
+        /// <para>jsonSerializer.NullValueHandling = NullValueHandling.Include.</para>
+        /// <para>jsonSerializer.ContractResolver = new DefaultContractResolver().</para>
+        /// </param>
+        public CosmosDbPartitionedStorage(CosmosDbPartitionedStorageOptions cosmosDbStorageOptions, JsonSerializer jsonSerializer)
+            : this(cosmosDbStorageOptions)
+        {
+            _jsonSerializer = jsonSerializer ?? throw new ArgumentNullException(nameof(jsonSerializer));
+        }
+
+        public async Task<IDictionary<string, object>> ReadAsync(string[] keys, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (keys == null)
+            {
+                throw new ArgumentNullException(nameof(keys));
+            }
+
+            if (keys.Length == 0)
+            {
+                // No keys passed in, no result to return.
+                return new Dictionary<string, object>();
+            }
+
+            // Ensure Initialization has been run
+            await InitializeAsync().ConfigureAwait(false);
+
+            var resultSetIterator = _container.GetItemQueryIterator<DocumentStoreItem>(
+                requestOptions: new QueryRequestOptions()
+                {
+                    PartitionKey = new PartitionKey(keys[0]),
+                });
+
+            var documentStoreItems = new List<DocumentStoreItem>(keys.Length);
+            while (resultSetIterator.HasMoreResults)
+            {
+                documentStoreItems.AddRange(await resultSetIterator.ReadNextAsync(cancellationToken).ConfigureAwait(false));
+            }
+
+            var storeItems = new Dictionary<string, object>(keys.Length);
+
+            foreach (var documentStoreItem in documentStoreItems)
```

Comment:
I believe this needs to be something more like:

```csharp
foreach (var documentStoreItem in documentStoreItems)
            {
                var item = documentStoreItem.Document.ToObject(typeof(object), _jsonSerializer);
                if (documentStoreItem is IStoreItem storeItem)
                {
                    storeItem.ETag = documentStoreItem.ETag;
                    storeItems.Add(documentStoreItem.RealId, item);
                }
            }
```

I'm getting `'Unable to cast object of type 'DocumentStoreItem' to type 'System.Collections.Generic.IDictionary`2[System.String,System.Object]'.'` down the line, otherwise.

Reply:
Done. Thanks.